aten/src/ATen/native/RowwisePrune.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#include <ATen/ATen.h>
+
+
+namespace at {
+namespace native {
+
+namespace {
+
+template <typename input_t>
+std::tuple<Tensor, Tensor> _rowwise_prune_helper(
+      const Tensor& weights, const Tensor& mask,
+      ScalarType compressed_indices_dtype) {
+  int num_non_masked_rows = 0;
+  auto mask_contig = mask.contiguous();
+  auto mask_data = mask_contig.data_ptr<bool>();
+  for (int i = 0; i < mask.numel(); ++i) {
+    num_non_masked_rows += (((mask_data[i] == true)) ? 1 : 0);
+  }
+  int num_cols = weights.size(1);
+  auto pruned_2d_tensor = at::empty({num_non_masked_rows, num_cols},
+      weights.options().dtype(at::kFloat));
+  auto* pruned_2d_tensor_data = pruned_2d_tensor.data_ptr<float>();
+  auto compressed_indices_mapping = at::empty({mask.numel()},
+      compressed_indices_dtype);
+  auto compressed_indices_mapping_data =
+      compressed_indices_mapping.data_ptr<input_t>();
+  auto weights_data = weights.data_ptr<float>();
+  int last_row_kept = 0;
+  for (int i = 0; i < mask.numel(); i++) {
+    if (mask_data[i]) {
+      memcpy(pruned_2d_tensor_data + last_row_kept * num_cols,
```

Comment:
Does this work for quantized weights?

Reply:
Good point. Made it generic to work all integral + float types using AT_DISPATCH_ALL_TYPES.